File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Node{
    int num;
    Node* left = NULL;
    Node* right = NULL;
};

Node* make_tree(int start, int end, vector<vector<int>>& nodes){
    if(start + 1 == end){
        Node* leaf = new Node();
        leaf->num = nodes[start][2];
        return leaf;
    }
    if(start + 1 > end) return NULL;
    
    int idx = max_element(nodes.begin() + start, nodes.begin() + end, [](vector<int> a, vector<int> b){ return a[1] < b[1]; }) - nodes.begin();
    
    Node* node = new Node();
    node->num = nodes[idx][2];
    node->left = make_tree(start, idx, nodes);
    node->right = make_tree(idx + 1, end, nodes);
    
    return node;
}

void preorder(Node* node, vector<int>& result){
    if(node == NULL) return;
    
    result.push_back(node->num);
    preorder(node->left, result);
    preorder(node->right, result);
}

void postorder(Node* node, vector<int>& result){
    if(node == NULL) return;
    
    postorder(node->left, result);
    postorder(node->right, result);
    result.push_back(node->num);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    int n = nodeinfo.size();
    for(int i = 0; i < nodeinfo.size(); i++) nodeinfo[i].push_back(i + 1);
    
    sort(nodeinfo.begin(), nodeinfo.end(), [](vector<int> a, vector<int> b){ return a[0] < b[0]; });
    
    Node* root = make_tree(0, n, nodeinfo);
    
    vector<int> pre;
    preorder(root, pre);
    
    vector<int> post;
    postorder(root, post);
    
    return {pre, post};
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys
sys.setrecursionlimit(10000)

class Node:
    def __init__(self, num, left, right):
        self.num = num
        self.left = left
        self.right = right

def make_tree(start, end, nodeinfo): # start <= i < end의 nodeinfo[i]로 tree를 만들어 가장 상위 노드를 반환
    if start >= end:
        return None
    
    idx = nodeinfo.index(max(nodeinfo[start:end], key=lambda n: n[1])) # 상위 노드의 index
    return Node(nodeinfo[idx][2], make_tree(start, idx, nodeinfo), make_tree(idx + 1, end, nodeinfo))

def preorder(node): # 전위 순회
    if node == None:
        return []
    
    return [node.num] + preorder(node.left) + preorder(node.right)

def postorder(node): # 후위 순회
    if node == None:
        return []
    
    return postorder(node.left) + postorder(node.right) + [node.num]

def solution(nodeinfo):
    for i in range(len(nodeinfo)):
        nodeinfo[i].append(i + 1) # 노드의 번호 적기
    
    nodeinfo.sort(key=lambda n: n[0]) # x좌표를 기준으로 오름차순 정렬
    
    root = make_tree(0, len(nodeinfo), nodeinfo)
    
    return [preorder(root), postorder(root)]